lib/offsite_payments/integrations/paytm.rb

```diff
@@ -0,0 +1,300 @@
+
+        check_sum = encrypted_data.delete("\n")
+        # new_pg_encrypt_variable(check_sum, key)
+        check_sum
+      end
+
+      class Helper < OffsitePayments::Helper
+        CHECKSUM_FIELDS = %w(MID ORDER_ID CUST_ID TXN_AMOUNT CHANNEL_ID INDUSTRY_TYPE_ID WEBSITE EMAIL MOBILE_NO).freeze
+
+        mapping :amount, 'TXN_AMOUNT'
+        mapping :account, 'MID'
+        mapping :order, 'ORDER_ID'
+        mapping :description, 'productinfo'
+
+        mapping :customer, first_name: 'firstname',
+                           last_name: 'lastname',
+                           email: 'CUST_ID',
+                           email: 'EMAIL',
+                           phone: 'MOBILE_NO'
+
+        mapping :billing_address, city: 'city',
+                                  address1: 'address1',
+                                  address2: 'address2',
+                                  state: 'state',
+                                  zip: 'zip',
+                                  country: 'country'
+
+        # Which tab you want to be open default on Paytm
+        # CC (CreditCard) or NB (NetBanking)
+        mapping :mode, 'pg'
+
+        mapping :notify_url, 'notify_url'
+        mapping :return_url, %w(surl furl)
+        mapping :cancel_return_url, 'curl'
+        mapping :checksum, 'hash'
+
+        mapping :user_defined, var1: 'udf1',
+                               var2: 'udf2',
+                               var3: 'udf3',
+                               var4: 'udf4',
+                               var5: 'udf5',
+                               var6: 'udf6',
+                               var7: 'udf7',
+                               var8: 'udf8',
+                               var9: 'udf9',
+                               var10: 'udf10'
+
+        def initialize(order, account, options = {})
+          super
+          @options = options
+          self.pg = 'CC'
+        end
+
+        def form_fields
+          sanitize_fields
+          @fields.merge(mappings[:CHECKSUMHASH] => generate_checksum)
+        end
+
+        def generate_checksum
+          @fields.merge(mappings[:CHANNEL_ID] => 'WEB')
+          @fields.merge(mappings[:INDUSTRY_TYPE_ID] => @options[:credential3])
+          @fields.merge(mappings[:WEBSITE] => @options[:credential4])
+          checksum_payload_items = CHECKSUM_FIELDS.map { |field| @fields[field] }
+          Paytm.checksum(checksum_payload_items, @options[:credential2])
+        end
+
+        def sanitize_fields
+          %w(address1 address2 city state country productinfo email phone).each do |field|
+            @fields[field].gsub!(/[^a-zA-Z0-9\-_@\/\s.]/, '') if @fields[field]
+          end
+        end
+      end
+
+      class Notification < OffsitePayments::Notification
+        PaytmResponseParams = %w(SUBS_ID MID BANKTXNID TXNAMOUNT CURRENCY STATUS RESPCODE RESPMSG TXNDATE GATEWAYNAME BANKNAME PAYMENTMODE PROMO_CAMP_ID PROMO_STATUS PROMO_RESPCODE ORDERID TXNID REFUNDAMOUNT REFID).freeze
```

Comment:
Constants have to be in the form of `PAYTM_RESPONSE_PARAMS` in Ruby. `PaytmResponseParams` can be misunderstood to be a class name.